timemachine/md/hrex.py

```diff
@@ -0,0 +1,247 @@
+
+
+@dataclass
+class Hrex(Generic[_Replica]):
+    replicas: List[_Replica]
+    replica_idx_by_state: List[ReplicaIdx]
+
+    @classmethod
+    def from_replicas(cls, replicas: Sequence[_Replica]) -> "Hrex":
+        return Hrex(list(replicas), [ReplicaIdx(i) for i, _ in enumerate(replicas)])
+
+    def sample_replicas(
+        self,
+        sample_replica: Callable[[_Replica, StateIdx], _Samples],
+        replica_from_samples: Callable[[_Samples], _Replica],
+    ) -> Tuple["Hrex[_Replica]", List[_Samples]]:
+
+        samples_by_state = [sample_replica(replica, state_idx) for state_idx, replica in self.state_replica_pairs]
+        replicas_by_state = [replica_from_samples(samples) for samples in samples_by_state]
+
+        replicas = list(self.replicas)
+        for state_idx, replica in enumerate(replicas_by_state):
+            replica_idx = self.replica_idx_by_state[state_idx]
+            replicas[replica_idx] = replica
+
+        return Hrex(replicas, self.replica_idx_by_state), samples_by_state
+
+    def attempt_neighbor_swaps(
+        self,
+        neighbor_pairs: Sequence[Tuple[StateIdx, StateIdx]],
+        log_q: Callable[[ReplicaIdx, StateIdx], float],
+        n_swap_attempts: int,
+    ) -> Tuple["Hrex[_Replica]", List[Tuple[int, int]]]:
+
+        move = MixtureOfMoves([NeighborSwapMove(log_q, s_a, s_b) for s_a, s_b in neighbor_pairs])
+
+        replica_idx_by_state = list(self.replica_idx_by_state)
+        for _ in range(n_swap_attempts):
+            replica_idx_by_state = move.move(replica_idx_by_state)
+
+        fraction_accepted_by_pair = list(zip(move.n_accepted_by_move, move.n_proposed_by_move))
+
+        return Hrex(self.replicas, replica_idx_by_state), fraction_accepted_by_pair
+
+    @property
+    def state_replica_pairs(self) -> List[Tuple[StateIdx, _Replica]]:
+        return [(StateIdx(i), self.replicas[replica_idx]) for i, replica_idx in enumerate(self.replica_idx_by_state)]
+
+
+def get_cumulative_replica_state_counts(replica_idx_by_state_by_iter: Sequence[Sequence[ReplicaIdx]]) -> NDArray:
+    """Given a mapping of state index to replica index by iteration, returns an array of cumulative counts by iteration, replica, and state.
+
+    Returns
+    -------
+    NDArray
+        (iter, replica, state) -> cumulative occupancy count: int
+    """
+    replica_idx_by_state_by_iter_ = np.array(replica_idx_by_state_by_iter)  # (iter, state) -> replica
+    _, n_states = replica_idx_by_state_by_iter_.shape
+    states = np.arange(n_states)
+    replica_in_state = replica_idx_by_state_by_iter_[:, :, None] == states  # (iter, replica, state) -> bool
+    cumulative_count = np.cumsum(replica_in_state.astype(int), axis=0)  # (iter, replica, state) -> int
+    return cumulative_count
+
+
+def estimate_transition_matrix(replica_idx_by_state_by_iter: Sequence[Sequence[ReplicaIdx]]) -> NDArray:
+    """Given a mapping of state index to replica index by iteration, returns an estimate of the transition matrix.
```

Comment:
Add documentation:

1) Square matrix but not expected to be symmetric
2) Markovian, but along which dimension? (i.e. reduction along which axis sums to 1?)

Reply:
Addressed in https://github.com/proteneer/timemachine/pull/1128/commits/8406c2164737ccc232d084094acd9fdb9b494af5 and https://github.com/proteneer/timemachine/pull/1128/commits/7994a93275dd5504af0dc901442fa664d73e56aa